module/zstd/zstd.c

```diff
@@ -0,0 +1,508 @@
+ * fields enclosed by brackets "[]" replaced with your own identifying
+ * information: Portions Copyright [yyyy] [name of copyright owner]
+ *
+ * CDDL HEADER END
+ */
+
+/*
+ * Copyright (c) 2016-2018 by Klara Systems Inc.
+ * Copyright (c) 2016-2018 Allan Jude <allanjude@freebsd.org>.
+ */
+
+#include <sys/param.h>
+#include <sys/sysmacros.h>
+#include <sys/zfs_context.h>
+#include <sys/zio_compress.h>
+#include <sys/spa.h>
+
+#define	ZSTD_STATIC_LINKING_ONLY
+#include <sys/zstd/zstd.h>
+#include <sys/zstd/zstd_errors.h>
+#include <sys/zstd/error_private.h>
+
+
+#define	ZSTD_KMEM_MAGIC		0x20160831
+
+/* for BSD compat */
+#define	__unused			__attribute__((unused))
+
+
+#ifdef _KERNEL
+#include <linux/sort.h>
+#define	zstd_qsort(a, n, es, cmp) sort(a, n, es, cmp, NULL)
+#else
+#define	zstd_qsort qsort
+#endif
+
+static size_t real_zstd_compress(const char *source, char *dest, int isize,
+    int osize, int level);
+static size_t real_zstd_decompress(const char *source, char *dest, int isize,
+    int maxosize);
+
+void *zstd_alloc(void *opaque, size_t size);
+void zstd_free(void *opaque, void *ptr);
+
+static const ZSTD_customMem zstd_malloc = {
+	zstd_alloc,
+	zstd_free,
+	NULL,
+};
+
+enum zstd_kmem_type {
+	ZSTD_KMEM_UNKNOWN = 0,
+	ZSTD_KMEM_CCTX,
+	ZSTD_KMEM_WRKSPC_4K_MIN,
+	ZSTD_KMEM_WRKSPC_4K_DEF,
+	ZSTD_KMEM_WRKSPC_4K_MAX,
+	ZSTD_KMEM_WRKSPC_16K_MIN,
+	ZSTD_KMEM_WRKSPC_16K_DEF,
+	ZSTD_KMEM_WRKSPC_16K_MAX,
+	ZSTD_KMEM_WRKSPC_128K_MIN,
+	ZSTD_KMEM_WRKSPC_128K_DEF,
+	ZSTD_KMEM_WRKSPC_128K_MAX,
+	/* SPA_MAXBLOCKSIZE */
+	ZSTD_KMEM_WRKSPC_MBS_MIN,
+	ZSTD_KMEM_WRKSPC_MBS_DEF,
+	ZSTD_KMEM_WRKSPC_MBS_MAX,
+	ZSTD_KMEM_DCTX,
+	ZSTD_KMEM_COUNT,
+};
+
+struct zstd_kmem {
+	uint_t			kmem_magic;
+	enum zstd_kmem_type	kmem_type;
+	size_t			kmem_size;
+};
+
+struct zstd_kmem_config {
+	size_t			block_size;
+	int			compress_level;
+	char			*cache_name;
+};
+
+static kmem_cache_t *zstd_kmem_cache[ZSTD_KMEM_COUNT] = { NULL };
+static struct zstd_kmem zstd_cache_size[ZSTD_KMEM_COUNT] = {
+	{ ZSTD_KMEM_MAGIC, 0, 0 } };
+static struct zstd_kmem_config zstd_cache_config[ZSTD_KMEM_COUNT] = {
+	{ 0, 0, "zstd_unknown" },
+	{ 0, 0, "zstd_cctx" },
+	{ 4096, ZIO_ZSTD_LEVEL_MIN, "zstd_wrkspc_4k_min" },
+	{ 4096, ZIO_ZSTD_LEVEL_DEFAULT, "zstd_wrkspc_4k_def" },
+	{ 4096, ZIO_ZSTD_LEVEL_MAX, "zstd_wrkspc_4k_max" },
+	{ 16384, ZIO_ZSTD_LEVEL_MIN, "zstd_wrkspc_16k_min" },
+	{ 16384, ZIO_ZSTD_LEVEL_DEFAULT, "zstd_wrkspc_16k_def" },
+	{ 16384, ZIO_ZSTD_LEVEL_MAX, "zstd_wrkspc_16k_max" },
+	{ SPA_OLD_MAXBLOCKSIZE, ZIO_ZSTD_LEVEL_MIN, "zstd_wrkspc_128k_min" },
+	{ SPA_OLD_MAXBLOCKSIZE, ZIO_ZSTD_LEVEL_DEFAULT,
+	    "zstd_wrkspc_128k_def" },
+	{ SPA_OLD_MAXBLOCKSIZE, ZIO_ZSTD_LEVEL_MAX, "zstd_wrkspc_128k_max" },
+	{ SPA_MAXBLOCKSIZE, ZIO_ZSTD_LEVEL_MIN, "zstd_wrkspc_mbs_min" },
+	{ SPA_MAXBLOCKSIZE, ZIO_ZSTD_LEVEL_DEFAULT, "zstd_wrkspc_mbs_def" },
+	{ SPA_MAXBLOCKSIZE, ZIO_ZSTD_LEVEL_MAX, "zstd_wrkspc_mbs_max" },
```

Comment:
Why not include the actual size in the same here like the others?  `16m`

Reply:
you mean naming SPA_MAXBLOCKSIZE to something like SPA_MAXBLOCKSIZE_16M? because it is 16m

Reply:
ah you mean the workspace name

Reply:
Whoops, yes.  That was a typo sorry, s/same/name/